### PR TITLE
New circular primitives: `Arc2d`, `CircularSector`, `CircularSegment`

### PR DESCRIPTION
**Pull request by spectria-limina**
## Objective

There are no 2d primitives available for the common shapes of an arc or a circular sector (pie slice).

## Solution

This PR introduces three new types to the existing math primitives:

- `Arc2d`: a portion of a (hollow) circle
- `CircularSector`: the area between an arc and the centre of the circle
- `CircularSegment`: the convex hull of an arc

Run `cargo run --example 2d_shapes` to see examples of the `CircularSector` and `CircularSegment` type.

## Design decisions

1. What should the arc type be named?

    **Answer:** `Arc2d`, because `Arc` is already taken by the standard library.

1. What should the origin point of these shapes be?
    
    - The origin point is the point around which the shape rotates, among other things.
    - For an arc and a circular sector, it's intuitive I think that the origin be the center of the corresponding circle.
    - But for a circular segment, I think the origin being the center of the chord is more intuitive, because it is easier to reason about rotation math that way.
    - On the other hand, consistency between the three shapes has value.
    
    **Answer:** This PR puts the origin of all three shapes in the center, prioritizing consistency, 

3. What should the UV-mapping of `CircularSector` and `CircularSegment` look like? There are two reasonable options:

    1. Treat the shapes as masks of the corresponding circle. Good for animating the texture, or cutting a single circular texture into pieces e.g. pizza slices.
    2. Map the shape's bounding box onto the texture. Good when the slice size is fixed and not going to change, to be maximally efficient with the texture space.
  
    **Answer:** This PR chooses option 1 as the default, but puts UV-mapping configuration in a `CircularMeshUvMode` enum, as futureproofing for later adding option 2.

5. How should the angle of these shapes be expressed?

    1. Follow mathematical conventions of starting from the right and going counterclockwise.
    2. Start from the top and extend by half the angle on either side.

    **Answer:** This PR chooses option 2, because it leads to better defaults and better ergonomics when reasoning about how to rotate the shape.

## TODO

- [x] Add convenience methods on `CircularSector` and `CircularSegment` that avoid needing to explicitly access the `arc` field.
- [x] Add examples that make use of textures, to verify that the UV mappings are correct.
- [x] Add testing, particularly of the shape math. (I've already caught at least one bug and there are probably more.) 

## Future Work

For after this PR is merged: 

- Add support for these shapes to `Gizmos` as in #11072.
- Add support for the alternative UV mode discussed above, as a configurable option on the `MeshBuilder` types.
- Integrate these shapes into the new `math/primitives` example.

## Changelog

- Added `Arc2d`, `CircularSector`, and `CircularSegment` primitives to [`bevy::math::primitives`].

### CODE REVIEW

File: examples/2d/2d_shapes.rs
Comment (Jondolf) [inline examples/2d/2d_shapes.rs:51]: This should be removed when we merge :)

File: crates/bevy_math/src/primitives/dim2.rs
Comment (Jondolf) [inline crates/bevy_math/src/primitives/dim2.rs:287]:
```suggestion
    /// If the arc is minor, i.e. less than half the circle, this will be the difference of the [`radius`](Self::radius)
```

**Review by Jondolf — COMMENTED** [review]
I didn't take a super in-depth look at all of the math and the bounding implementation yet, but everything seems correct at a glance. Code quality overall is great!

I mainly left some documentation and naming suggestions and typo fixes.

As for the unresolved questions:

1. I think the origin should probably be at the circle's center for all three shapes for the sake of consistency. I can't think of another origin that would make much sence, except maybe for the circular segment, but I think I prefer consistency here. Whatever is chosen, the important thing is that it is documented (and it already is).
2. Option 1 (the current one) is a good default. We can add option 2 as a configuration option in a follow-up.
3. I agree, a half-angle in radians extending symmetrically in both directions from the vertical (i.e. the "top" of the circle) might be more intuitive and easier to work with. Starting from the right and going counterclockwise might be more conventional mathematically (though I'm not sure if arcs specifically have a convention for this), but I feel like that wouldn't be a very useful default from a practical standpoint. I don't have a strong opinion on this though, so other peoples' thoughts on the topic would be useful.

File: crates/bevy_math/src/primitives/dim2.rs
Comment (Jondolf) [inline crates/bevy_math/src/primitives/dim2.rs:359]:
```suggestion
    #[cfg_attr(feature = "serialize", serde(flatten))]
```

File: examples/2d/2d_shapes.rs
Comment (spectria-limina) [reply examples/2d/2d_shapes.rs]: Aww...

**Comment by spectria-limina** [comment]
Thanks for the review! I've made the changes plus a few more things. I'll update the first comment to reflect the current state of things.

File: crates/bevy_math/src/bounding/bounded2d/primitive_impls.rs
Comment (alice-i-cecile) [inline crates/bevy_math/src/bounding/bounded2d/primitive_impls.rs:99]:
```suggestion
        // Otherwise, we use the same logic as for Arc2d, above, just with the circle's center as an additional possibility.
```

File: crates/bevy_math/src/primitives/dim2.rs
Comment (alice-i-cecile) [inline crates/bevy_math/src/primitives/dim2.rs:74]:
```suggestion
/// use the pie-shaped [`CircularSector`].
```

File: crates/bevy_math/src/primitives/dim2.rs
Comment (alice-i-cecile) [inline crates/bevy_math/src/primitives/dim2.rs:76]:
```suggestion
/// use the bowl-shaped [`CircularSegment`].
```

File: crates/bevy_math/src/primitives/dim2.rs
Comment (alice-i-cecile) [inline crates/bevy_math/src/primitives/dim2.rs:82]: Hmm. Why do you think we should allow these be constructed, rather than enforcing this via the API?

File: crates/bevy_math/src/primitives/dim2.rs
Comment (alice-i-cecile) [inline crates/bevy_math/src/primitives/dim2.rs:254]: This needs clarification in less-jargony terms.

File: crates/bevy_render/src/mesh/primitives/dim2.rs
Comment (alice-i-cecile) [inline crates/bevy_render/src/mesh/primitives/dim2.rs:87]: This is a little odd as an enum with only one variant. I would like a comment here recording how it might be extended.

File: examples/2d/2d_shapes.rs
Comment (Jondolf) [inline examples/2d/2d_shapes.rs:26]: Having structs inside functions is a bit unconventional imo; I'd move this above `Setup`

File: Cargo.toml
Comment (Jondolf) [inline Cargo.toml:406]: I'm wondering if there's a better name... "circular" might imply that this also includes circles and maybe ellipses, but it just has the arc-like shapes. `mesh2d_circular` and `mesh2d_arcs` would both be fine probably

File: examples/2d/mesh2d_circular.rs
Comment (Jondolf) [inline examples/2d/mesh2d_circular.rs:121]: Is this intended to stay here? I'm guessing it was just for debugging and can be removed
```suggestion
```

**Review by Jondolf — CHANGES_REQUESTED** [review]
I love the tests and overall code quality, and the examples looks great. Large PR though!

I left a ton of tweaks and suggestions regarding names, docs, and some other things. I tried to be pretty thorough, so apologies if some things are overly nitpicky :)

File: Cargo.toml
Comment (spectria-limina) [reply Cargo.toml:413]: I like `_arcs` better.

File: crates/bevy_math/src/primitives/dim2.rs
Comment (spectria-limina) [reply crates/bevy_math/src/primitives/dim2.rs:82]: I don't feel strongly, but it's consistent with the other primitives to expose the fields and thereby let you construct invalid ones.

File: crates/bevy_math/src/primitives/dim2.rs
Comment (spectria-limina) [reply crates/bevy_math/src/primitives/dim2.rs:68]: I'm going to have to update a lot of these...

File: crates/bevy_math/src/primitives/dim2.rs
Comment (spectria-limina) [reply crates/bevy_math/src/primitives/dim2.rs:90]: I went with "defining", as I think it's probably pretty clear that an arc is defined by an angle.

File: examples/README.md
Comment (Jondolf) [inline examples/README.md:102]: This should match the Cargo.toml
```suggestion
[Arc 2D Meshes](../examples/2d/mesh2d_arcs.rs) | Demonstrates UV-mapping of the circular segment and sector primitives
```

File: Cargo.toml
Comment (Jondolf) [inline Cargo.toml:411]: Forgot to update this too
```suggestion
[package.metadata.example.mesh2d_arcs]
```

**Review by Jondolf — APPROVED** [review]
Alright, once the example metadata is fixed, I think this LGTM! Thanks for the patience :)

File: crates/bevy_math/src/primitives/dim2.rs
Comment (spectria-limina) [reply crates/bevy_math/src/primitives/dim2.rs:134]: Turns is better.

File: crates/bevy_math/src/primitives/dim2.rs
Comment (spectria-limina) [reply crates/bevy_math/src/primitives/dim2.rs:359]: I think the link to `Arc2d::sagitta` is enough?

**Comment by spectria-limina** [comment]
I've now updated the PR and got it as close to working as I can, but #12729 is preventing all tests from passing and I can't figure out a fix locally.

File: crates/bevy_math/src/primitives/dim2.rs
Comment (spectria-limina) [reply crates/bevy_math/src/primitives/dim2.rs:338]: As below, I think the link is fine.

**Review by Jondolf — APPROVED** [review]

**Comment by bugsweeper** [comment]
> There also seems to be a failing wasm build that isn't related to my PR.

That has been fixed #12730

**Review by mweatherley — APPROVED** [review]
I have a couple of minor nits, but I would like for this to get merged. I wish I had realized this needed another review earlier; I'm sorry that this PR has been neglected somewhat.

Overall, I am very happy with the documentation, testing, and examples.

**Comment by alice-i-cecile** [comment]
@spectria-limina I'm happy to merge this as is: let me know if you're able to resolve merge conflicts?

**Pull request closed without merging**
